src/config/format.rs

```diff
@@ -0,0 +1,258 @@
+//! Support for loading configs from multiple formats.
+
+#![deny(missing_docs, missing_debug_implementations)]
+
+use serde::de;
+use std::path::Path;
+
+/// The format used to represent the configuration data.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub enum Format {
+    /// The format could not be determined.
+    Unknown,
```

Comment:
Since the fallback is TOML, do you think this is necessary?

Reply:
It allows us to iteratively try and detect the format in multiple steps, and we only fallback at the very end.

There was a discussion about detecting the format based on the contents - I think it will be very useful for that case. Or for other extensions to this format detection process.

Reply:
I do not see anything in #3174 about supporting files which are not postfixed with `.json`, `.toml`, or `.yaml`.

In that issue @lukesteensen even requested a flag to configure when json was parsed. In that issue yaml was not discussed at all. While I do not think a flag is neccessary, or that supporting yaml is a bad idea, those are discussions that should be had before a merge. :)

Reply:
Yep, I'm fine with this! Let's move this discussion to the regular PR comments stream then.

Reply:
I think we discussed removing this variant?

Reply:
We talked about generating an error when the format is unknown - and this I didn't do yet to preserve backward compatibility. We can do this in 0.12.
Removing this variant is a bit different - it would be eliminating the whole notion of delayed format resolution (unless we switch to `Option<Format>` which is also fine for the job) - which I think was also discussed, and I think we settled on keeping the overall idea. Or didn't we? Sorry if I've missed anything.

Reply:
I also would prefer `Option<Format>`. It's easier to reason about if in all cases we resolved/assigned default format.